gateway/ln-gateway/src/lightning/ldk.rs

```diff
@@ -0,0 +1,501 @@
+use std::path::Path;
+use std::str::FromStr;
+use std::sync::Arc;
+use std::time::Duration;
+
+use async_trait::async_trait;
+use bitcoin::{secp256k1, Network, OutPoint};
+use fedimint_core::runtime::spawn;
+use fedimint_core::task::TaskGroup;
+use ldk_node::lightning::ln::msgs::SocketAddress;
+use ldk_node::lightning::ln::PaymentHash;
+use ldk_node::lightning_invoice::Bolt11Invoice;
+use ldk_node::payment::{PaymentKind, PaymentStatus};
+use lightning::ln::PaymentPreimage;
+use tokio::sync::mpsc::Sender;
+use tokio_stream::wrappers::ReceiverStream;
+use tonic::Status;
+use tracing::{error, info};
+
+use super::cln::RouteHtlcStream;
+use super::{ChannelInfo, ILnRpcClient, LightningRpcError};
+use crate::gateway_lnrpc::create_invoice_request::Description;
+use crate::gateway_lnrpc::intercept_htlc_response::{Action, Settle};
+use crate::gateway_lnrpc::{
+    CloseChannelsWithPeerResponse, CreateInvoiceRequest, CreateInvoiceResponse, EmptyResponse,
+    GetFundingAddressResponse, GetNodeInfoResponse, GetRouteHintsResponse, InterceptHtlcRequest,
+    InterceptHtlcResponse, PayInvoiceRequest, PayInvoiceResponse,
+};
+
+pub struct GatewayLdkClient {
+    /// The underlying lightning node.
+    node: Arc<ldk_node::Node>,
+
+    /// The client for querying data about the blockchain.
+    esplora_client: esplora_client::AsyncClient,
+
+    /// A handle to the task that seeds the HTLC stream.
+    /// TODO: This should be a shutdown sender instead, and we can discard the
+    /// handle.
+    htlc_stream_seeder_task_handle: tokio::task::JoinHandle<()>,
+
+    /// The HTLC stream, until it is taken by calling
+    /// `ILnRpcClient::route_htlcs`.
+    htlc_stream_receiver_or:
+        Option<tokio::sync::mpsc::Receiver<Result<InterceptHtlcRequest, Status>>>,
```

Comment:
This seems really awkward to me. Does `GatewayLdkClient` need to own this? Can it be passed in as an argument to a function?

Reply:
The gateway lightning client can be in two states - either it's inactive (not intercepting HTLCs) or active (intercepting HTLCs), and calling `ILnRpcClient::route_htlcs` spins up the HTLC interception, switching from the former state to the latter. Currently these two states are internally represented in the following way:

**Inactive**: `GatewayLdkClient.htlc_stream_seeder_task_handle` is already running and passing incoming HTLCs to the HTLC stream receiver. The HTLC stream receiver is held in `GatewayLdkClient.htlc_stream_receiver_or` until `route_htlcs` is called, at which point this receiver is taken and returned.

**Active**: `GatewayLdkClient.htlc_stream_seeder_task_handle` is still running. But the HTLC stream receiver has been passed to the caller of `route_htlcs` and so `GatewayLdkClient.htlc_stream_receiver_or` is `None`.

What if we changed `GatewayLdkClient.htlc_stream_seeder_task_handle` to be an `Option<tokio::task::JoinHandle<()>>`, removed `GatewayLdkClient.htlc_stream_receiver_or` altogether, and represented the states like this:

**Inactive**: `GatewayLdkClient.htlc_stream_seeder_task_handle_or` is `None` and `ldk-node` events are not being handled yet. The HTLC stream receiver doesn't exist yet.

**Active**: The HTLC stream sender and receiver are created when `route_htlcs` is called. The sender is passed into the htlc stream seeder task, which is started up and assigned to `GatewayLdkClient.htlc_stream_seeder_task_handle_or`. The receiver is returned by `route_htlcs` and never held by `GatewayLdkClient`.

WDYT?

Reply:
Let's chat about this offline